src/loader/bgw_counter.c

```diff
@@ -98,69 +98,54 @@ bgw_counter_reinit(void)
 }
 
 extern bool
-bgw_total_workers_increment()
+bgw_total_workers_increment_by(int increment_by)
 {
 	bool		incremented = false;
 	int			max_workers = guc_max_background_workers;
 
 	SpinLockAcquire(&ct->mutex);
-	if (ct->total_workers < max_workers)
+	if (ct->total_workers + increment_by <= max_workers)
 	{
-		ct->total_workers++;
+		ct->total_workers += increment_by;
 		incremented = true;
 	}
 	SpinLockRelease(&ct->mutex);
 	return incremented;
 }
 
-extern void
-bgw_total_workers_decrement()
-{
-	/*
-	 * Launcher is 1 worker, and when it dies we reinitialize, so we should
-	 * never be below 1
-	 */
-	SpinLockAcquire(&ct->mutex);
-	if (ct->total_workers > 1)
-		ct->total_workers--;
-	else
-		ereport(ERROR, (errmsg("TimescaleDB background worker internal error: cannot decrement workers below 1"),
-						errhint("The background worker scheduler is in an invalid state, please submit a bug report.")));
-	SpinLockRelease(&ct->mutex);
-}
-
 extern bool
-bgw_total_workers_increment_by(int increment_by)
+bgw_total_workers_increment()
 {
-	bool		incremented = false;
-	int			max_workers = guc_max_background_workers;
-
-	SpinLockAcquire(&ct->mutex);
-	if (ct->total_workers + increment_by <= max_workers)
-	{
-		ct->total_workers += increment_by;
-		incremented = true;
-	}
-	SpinLockRelease(&ct->mutex);
-	return incremented;
+	return bgw_total_workers_increment_by(1);
 }
 
 extern void
 bgw_total_workers_decrement_by(int decrement_by)
 {
-	SpinLockAcquire(&ct->mutex);
-
 	/*
 	 * Launcher is 1 worker, and when it dies we reinitialize, so we should
 	 * never be below 1
 	 */
+	SpinLockAcquire(&ct->mutex);
 	if (ct->total_workers - decrement_by >= 1)
+	{
 		ct->total_workers -= decrement_by;
+		SpinLockRelease(&ct->mutex);
+	}
 	else
-		ereport(ERROR, (errmsg("TimescaleDB background worker internal error: cannot decrement workers below 1"),
-						errhint("The background worker scheduler is in an invalid state, please submit a bug report.")));
-	SpinLockRelease(&ct->mutex);
+	{
+		SpinLockRelease(&ct->mutex);
+		ereport(LOG, (errmsg("TimescaleDB background worker cannot decrement workers below 1"),
```

Comment:
It is unclear, and somewhat inconsistent, that we do not return success or failure here like we do for increment. 

I also question the usefulness of logging the invalid state here, because it signals we don't have a handle on what is going on in our own code. 

If trying to decrement too much is something that should never occur, shouldn't we `Assert()` it and test it so that we are confident that it works as it should? A hard fail like before is probably better, although my original comment about not doing this in these types of functions stands. (Thus returning success or failure makes sense.)

Otherwise, what is the bar for adding this kind of invalid state log statements? We could have these littered across our code eventually, and that doesn't make much sense IMO.